### PR TITLE
[BUGFIX] Ensure that dynamic injections are not cacheable in a Container.

### PR DESCRIPTION
**Pull request by dgeb**
A non-singleton injection should prevent caching of an object or factory instance in a container.

Fixes #3310

### CODE REVIEW

**Comment by rwjblue** [comment]
LGTM, @stefanpenner  - r?

File: packages/container/lib/container.js [hunk 1]
```diff
@@ -182,6 +182,10 @@ Container.prototype = {
   }
 })();
 
+function isSingleton(container, fullName) {
+  return container._registry.getOption(fullName, 'singleton') !== false;
```
Comment (twokul) [inline packages/container/lib/container.js]: `return !container._registry.getOption(fullName, 'singleton')` perhaps?

Comment (twokul) [inline packages/container/lib/container.js]: or even `return container._registry.getOption(fullName, 'singleton')` if I'm reading this code correctly.

Comment (dgeb) [inline packages/container/lib/container.js]: The `singleton` option by default is `true`, so neither alternative above would work for all cases (`true`, `false`, `undefined`). One alternative that would work is `return !!container._registry.getOption(fullName, 'singleton');` - I'm guessing that's what you meant with your first suggestion. Anyway, I'd be glad to change to this if it's more a more common convention throughout ember. There are other instances of checking `!== false` throughout the Container which I could also change to `!!` if there's consensus on this.

Comment (dgeb) [inline packages/container/lib/container.js]: On second thought, I don't see an upside to being more accepting of truthy / falsy values in settings completely controlled by developers (which should therefore have explicit values like `false`), so I'd recommend keeping it as is.

Comment (rwjblue) [inline packages/container/lib/container.js]: Agreed. Let's keep it matching the exact value specified.

**Comment by rwjblue** [comment]
@dgeb - This looks good, but needs a rebase.  Can you confirm this is still needed (after the container/registry reform PR's) and rebase ?

**Comment by rwjblue** [comment]
@dgeb - Ping?

**Comment by dgeb** [comment]
@rwjblue I can confirm this is still needed via the test that's included, which fails if injections aren't marked as dynamic.

I've just rebased against master.

**Comment by rwjblue** [comment]
Thank you @dgeb!

**Pull request merged**